src/modules/audio/spatializers/oculus_spatializer.c

```diff
@@ -0,0 +1,222 @@
+#include "audio/audio.h"
+#include "audio/spatializer.h"
+#include "lib/miniaudio/miniaudio.h"
+#include "oculus_spatializer_math_shim.h"
+#include "OVR_Audio.h"
+#include <stdlib.h>
+
+typedef struct {
+  Source* source;
+  bool usedSourceThisPlayback; // If true source was non-NULL at some point between midPlayback going high and tail()
+  bool occupied; // If true either source->playing or Oculus Audio is doing an echo tailoff
+} SourceRecord;
+
+struct {
+  uint32_t sampleRate;
+
+  ovrAudioContext context;
+  SourceRecord *sources;
+  int sourceMax;   // Maximum allowed simultaneous sources
+
+  int sourceCount; // Number of active sources seen this playback
+  int occupiedCount; // Number of sources+tailoffs seen this playback (ie strictly gte sourceCount)
+  bool midPlayback; // An onPlayback callback is in progress
+
+  bool poseUpdated; // setListenerPose has been called since the last playback
+  ovrPoseStatef pose;
+  ma_mutex poseLock; // Using ma_mutex in case holding a lovr lock inside a ma lock is weird
+  bool poseLockInited;
+} state;
+
+static bool oculus_spatializer_init(SpatializerConfigIn configIn, SpatializerConfigOut *configOut)
+{
+  // Initialize own state
+  state.sampleRate = configIn.sampleRate;
+  configOut->needFixedBuffer = true;
+  state.sourceMax = configIn.maxSourcesHint;
+  state.sources = calloc(state.sourceMax, sizeof(SourceRecord));
+
+  if (!state.poseLockInited) {
+    int mutexStatus = ma_mutex_init(&state.poseLock);
+    lovrAssert(mutexStatus == MA_SUCCESS, "Failed to create audio mutex");
+    state.poseLockInited = true;
+  }

```

Comment:
I don't kill the mutex between destroy/init cycles. Is this the correct convention? I'm terrified of deleting a mutex while a thread is waiting on it

Reply:
It seems like it should be destroyed somewhere.  I think we can maintain invariants that A) the callbacks aren't firing when the spatializer is destroyed (devices aren't started while spatializer is being created/destroyed) B) Sources aren't doing anything while spatializers are being created/destroyed (this requires spatializers to only be destroyed in lovrAudioDestroy after sources are gone and/or destroying a spatializer any other time to take out the playback lock maybe?)